diff --git a/src/colmap/mvs/model.h b/src/colmap/mvs/model.h
index 15761a94..8f6d9fc7 100644
--- a/src/colmap/mvs/model.h
+++ b/src/colmap/mvs/model.h
@@ -50,9 +50,9 @@ namespace mvs {
 // Simple sparse model class.
 struct Model {
   struct Point {
-    float x = 0;
-    float y = 0;
-    float z = 0;
+    double x = 0;
+    double y = 0;
+    double z = 0;
     std::vector<int> track;
   };
 
@@ -77,8 +77,9 @@ struct Model {
   const std::vector<std::vector<int>>& GetMaxOverlappingImagesFromPMVS() const;
 
   // Compute the robust minimum and maximum depths from the sparse point cloud.
-  std::vector<std::pair<float, float>> ComputeDepthRanges() const;
-
+  std::vector<std::pair<float, float>> GetDepthRanges() const {
+    return depth_ranges_;
+  }
   // Compute the number of shared points between all overlapping images.
   std::vector<std::map<int, int>> ComputeSharedPoints() const;
 
@@ -102,6 +103,7 @@ struct Model {
   std::unordered_map<std::string, int> image_name_to_idx_;
 
   std::vector<std::vector<int>> pmvs_vis_dat_;
+  std::vector<std::pair<float, float>> depth_ranges_;
 };
 
 }  // namespace mvs
