diff --git a/src/colmap/ui/bundle_adjustment_widget.cc b/src/colmap/ui/bundle_adjustment_widget.cc
index d23e36c0..e86a5cbb 100644
--- a/src/colmap/ui/bundle_adjustment_widget.cc
+++ b/src/colmap/ui/bundle_adjustment_widget.cc
@@ -98,7 +98,7 @@ void BundleAdjustmentWidget::Run() {
 
   // Normalize scene for numerical stability and
   // to avoid large scale changes in viewer.
-  reconstruction_->Normalize();
+  // reconstruction_->Normalize();
 
   thread_control_widget_->StartThread("Bundle adjusting...", true,
                                       std::move(thread));
