diff --git a/src/colmap/estimators/bundle_adjustment.cc b/src/colmap/estimators/bundle_adjustment.cc
index 5ac35f53..a97fcb9b 100644
--- a/src/colmap/estimators/bundle_adjustment.cc
+++ b/src/colmap/estimators/bundle_adjustment.cc
@@ -1,3 +1,35 @@
+// ===============================================================================================================
+// Copyright (c) 2019, Cornell University. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that
+// the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright otice, this list of conditions and
+//       the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and
+//       the following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of Cornell University nor the names of its contributors may be used to endorse or
+//       promote products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+// TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+// OF SUCH DAMAGE.
+//
+// Author: Kai Zhang (kz298@cornell.edu)
+//
+// The research is based upon work supported by the Office of the Director of National Intelligence (ODNI),
+// Intelligence Advanced Research Projects Activity (IARPA), via DOI/IBC Contract Number D17PC00287.
+// The U.S. Government is authorized to reproduce and distribute copies of this work for Governmental purposes.
+// ===============================================================================================================
+//
+//
 // Copyright (c) 2023, ETH Zurich and UNC Chapel Hill.
 // All rights reserved.
 //
@@ -66,8 +98,29 @@ ceres::LossFunction* BundleAdjustmentOptions::CreateLossFunction() const {
   return loss_function;
 }
 
+ceres::LossFunction* BundleAdjustmentOptions::CreateConstrainPointsLossFunction() const {
+    ceres::LossFunction* loss_function = nullptr;
+    switch (constrain_points_loss) {
+      case LossFunctionType::TRIVIAL:
+        loss_function = new ceres::ScaledLoss(new ceres::TrivialLoss(), constrain_points_loss_weight,
+            ceres::TAKE_OWNERSHIP);
+        break;
+      case LossFunctionType::SOFT_L1:
+        loss_function = new ceres::ScaledLoss(new ceres::SoftLOneLoss(constrain_points_loss_scale), constrain_points_loss_weight,
+            ceres::TAKE_OWNERSHIP);
+        break;
+      case LossFunctionType::CAUCHY:
+        loss_function = new ceres::ScaledLoss(new ceres::CauchyLoss(constrain_points_loss_scale), constrain_points_loss_weight,
+            ceres::TAKE_OWNERSHIP);
+        break;
+    }
+    CHECK_NOTNULL(loss_function);
+    return loss_function;
+}
+
 bool BundleAdjustmentOptions::Check() const {
   CHECK_OPTION_GE(loss_function_scale, 0);
+  CHECK_OPTION_GE(constrain_points_loss_scale, 0);
   CHECK_OPTION_LT(max_num_images_direct_dense_cpu_solver,
                   max_num_images_direct_sparse_cpu_solver);
   CHECK_OPTION_LT(max_num_images_direct_dense_gpu_solver,
@@ -222,6 +275,10 @@ const std::unordered_set<point3D_t>& BundleAdjustmentConfig::ConstantPoints()
   return constant_point3D_ids_;
 }
 
+const std::unordered_set<point3D_t>& BundleAdjustmentConfig::ConstrainedPoints() const {
+  return constrained_point3D_ids_;
+}
+
 const std::unordered_set<image_t>& BundleAdjustmentConfig::ConstantCamPoses()
     const {
   return constant_cam_poses_;
@@ -242,6 +299,11 @@ void BundleAdjustmentConfig::AddConstantPoint(const point3D_t point3D_id) {
   constant_point3D_ids_.insert(point3D_id);
 }
 
+void BundleAdjustmentConfig::AddConstrainedPoint(const point3D_t point3D_id) {
+   CHECK(!HasConstrainedPoint(point3D_id));
+   constrained_point3D_ids_.insert(point3D_id);
+}
+
 bool BundleAdjustmentConfig::HasPoint(const point3D_t point3D_id) const {
   return HasVariablePoint(point3D_id) || HasConstantPoint(point3D_id);
 }
@@ -256,6 +318,10 @@ bool BundleAdjustmentConfig::HasConstantPoint(
   return constant_point3D_ids_.find(point3D_id) != constant_point3D_ids_.end();
 }
 
+bool BundleAdjustmentConfig::HasConstrainedPoint(const point3D_t point3D_id) const {
+   return constrained_point3D_ids_.find(point3D_id) != constrained_point3D_ids_.end();
+}
+
 void BundleAdjustmentConfig::RemoveVariablePoint(const point3D_t point3D_id) {
   variable_point3D_ids_.erase(point3D_id);
 }
@@ -264,6 +330,10 @@ void BundleAdjustmentConfig::RemoveConstantPoint(const point3D_t point3D_id) {
   constant_point3D_ids_.erase(point3D_id);
 }
 
+void BundleAdjustmentConfig::RemoveConstrainedPoint(const point3D_t point3D_id) {
+  constrained_point3D_ids_.erase(point3D_id);
+}
+
 ////////////////////////////////////////////////////////////////////////////////
 // BundleAdjuster
 ////////////////////////////////////////////////////////////////////////////////
@@ -329,6 +399,12 @@ void BundleAdjuster::SetUpProblem(Reconstruction* reconstruction,
     AddPointToProblem(point3D_id, reconstruction, loss_function);
   }
 
+  // add constrained points
+  ceres::LossFunction* constrain_points_loss_function = options_.CreateConstrainPointsLossFunction();
+  for (const auto point3D_id : config_.ConstrainedPoints()) {
+    AddConstrainedPointToProblem(point3D_id, reconstruction, constrain_points_loss_function);
+  }
+
   ParameterizeCameras(reconstruction);
   ParameterizePoints(reconstruction);
 }
@@ -514,6 +590,16 @@ void BundleAdjuster::AddPointToProblem(const point3D_t point3D_id,
   }
 }
 
+void BundleAdjuster::AddConstrainedPointToProblem(const point3D_t point3D_id,
+                                       Reconstruction* reconstruction,
+                                       ceres::LossFunction* loss_function) {
+  Point3D& point3D = reconstruction->Point3D(point3D_id);
+
+  ceres::CostFunction* cost_function = Point3DCostFunction::Create(point3D.xyz);
+
+  problem_->AddResidualBlock(cost_function, loss_function, point3D.xyz.data());
+}
+
 void BundleAdjuster::ParameterizeCameras(Reconstruction* reconstruction) {
   const bool constant_camera = !options_.refine_focal_length &&
                                !options_.refine_principal_point &&
