diff --git a/src/colmap/ui/bundle_adjustment_widget.cc b/src/colmap/ui/bundle_adjustment_widget.cc
index fe6143c9..498bd448 100644
--- a/src/colmap/ui/bundle_adjustment_widget.cc
+++ b/src/colmap/ui/bundle_adjustment_widget.cc
@@ -108,7 +108,7 @@ void BundleAdjustmentWidget::Run() {
 
   // Normalize scene for numerical stability and
   // to avoid large scale changes in viewer.
-  reconstruction_->Normalize();
+  // reconstruction_->Normalize();
 
   thread_control_widget_->StartThread(
       "Bundle adjusting...", true, std::move(thread));
