diff --git a/src/colmap/exe/sfm.h b/src/colmap/exe/sfm.h
index 56329055..914b5763 100644
--- a/src/colmap/exe/sfm.h
+++ b/src/colmap/exe/sfm.h
@@ -46,6 +46,8 @@ int RunPointTriangulatorImpl(
 
 int RunAutomaticReconstructor(int argc, char** argv);
 int RunBundleAdjuster(int argc, char** argv);
+// @Sebastian
+int RunNormalize(int argc, char** argv);
 int RunColorExtractor(int argc, char** argv);
 int RunMapper(int argc, char** argv);
 int RunHierarchicalMapper(int argc, char** argv);
