diff --git a/src/exe/sfm.h b/src/exe/sfm.h
index 56bb07c5..ab2af45a 100644
--- a/src/exe/sfm.h
+++ b/src/exe/sfm.h
@@ -46,6 +46,8 @@ int RunPointTriangulatorImpl(Reconstruction& reconstruction,
 
 int RunAutomaticReconstructor(int argc, char** argv);
 int RunBundleAdjuster(int argc, char** argv);
+// @Sebastian
+int RunNormalize(int argc, char** argv);
 int RunColorExtractor(int argc, char** argv);
 int RunMapper(int argc, char** argv);
 int RunHierarchicalMapper(int argc, char** argv);
