diff --git a/src/colmap/exe/colmap.cc b/src/colmap/exe/colmap.cc
index 238eee4d..ea602dbb 100644
--- a/src/colmap/exe/colmap.cc
+++ b/src/colmap/exe/colmap.cc
@@ -140,6 +140,9 @@ int main(int argc, char** argv) {
   commands.emplace_back("vocab_tree_matcher", &colmap::RunVocabTreeMatcher);
   commands.emplace_back("vocab_tree_retriever", &colmap::RunVocabTreeRetriever);
 
+  // @kai
+  commands.emplace_back("normalize", &colmap::RunNormalize);
+
   if (argc == 1) {
     return ShowHelp(commands);
   }
