diff --git a/src/exe/colmap.cc b/src/exe/colmap.cc
index 050a3a0d..ff7e8f66 100644
--- a/src/exe/colmap.cc
+++ b/src/exe/colmap.cc
@@ -144,6 +144,9 @@ int main(int argc, char** argv) {
   commands.emplace_back("vocab_tree_matcher", &RunVocabTreeMatcher);
   commands.emplace_back("vocab_tree_retriever", &RunVocabTreeRetriever);
 
+  // @kai
+  commands.emplace_back("normalize", &RunNormalize);
+
   if (argc == 1) {
     return ShowHelp(commands);
   }
