diff --git a/src/colmap/util/option_manager.cc b/src/colmap/util/option_manager.cc
index 1a887604..6f11a0d7 100644
--- a/src/colmap/util/option_manager.cc
+++ b/src/colmap/util/option_manager.cc
@@ -1,3 +1,35 @@
+// ===============================================================================================================
+// Copyright (c) 2019, Cornell University. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that
+// the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright otice, this list of conditions and
+//       the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and
+//       the following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of Cornell University nor the names of its contributors may be used to endorse or
+//       promote products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+// TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+// OF SUCH DAMAGE.
+//
+// Author: Kai Zhang (kz298@cornell.edu)
+//
+// The research is based upon work supported by the Office of the Director of National Intelligence (ODNI),
+// Intelligence Advanced Research Projects Activity (IARPA), via DOI/IBC Contract Number D17PC00287.
+// The U.S. Government is authorized to reproduce and distribute copies of this work for Governmental purposes.
+// ===============================================================================================================
+//
+//
 // Copyright (c) 2023, ETH Zurich and UNC Chapel Hill.
 // All rights reserved.
 //
@@ -472,6 +504,15 @@ void OptionManager::AddBundleAdjustmentOptions() {
                               &bundle_adjustment->refine_extra_params);
   AddAndRegisterDefaultOption("BundleAdjustment.refine_extrinsics",
                               &bundle_adjustment->refine_extrinsics);
+  // constrain points options
+  AddAndRegisterDefaultOption("BundleAdjustment.constrain_points",
+                              &bundle_adjustment->constrain_points);
+  //  AddAndRegisterDefaultOption("BundleAdjustment.constrain_points_loss",
+  //                              &bundle_adjustment->constrain_points_loss);
+  AddAndRegisterDefaultOption("BundleAdjustment.constrain_points_loss_scale",
+                              &bundle_adjustment->constrain_points_loss_scale);
+  AddAndRegisterDefaultOption("BundleAdjustment.constrain_points_loss_weight",
+                              &bundle_adjustment->constrain_points_loss_weight);
 }
 
 void OptionManager::AddMapperOptions() {
@@ -651,6 +692,11 @@ void OptionManager::AddPatchMatchStereoOptions() {
                               &patch_match_stereo->allow_missing_files);
   AddAndRegisterDefaultOption("PatchMatchStereo.write_consistency_graph",
                               &patch_match_stereo->write_consistency_graph);
+  // whether to overwrite existing reconstruction
+  AddAndRegisterDefaultOption("PatchMatchStereo.overwrite",
+                              &patch_match_stereo->overwrite);
+  AddAndRegisterDefaultOption("PatchMatchStereo.use_exist_photom",
+                              &patch_match_stereo->use_exist_photom);
 }
 
 void OptionManager::AddStereoFusionOptions() {
